File: codeforces/MaratonaMineira2022/D/D.py
import sys

MAX = 1000000001

def fill_prefix_sum(country: list[list[int]], t: int) -> list[list[int]]:
    n, m = len(country), len(country[0])
    prefix_sum = [[0 for _ in range(m)] for _ in range(n)]

    prefix_sum[0][0] = int(country[0][0] <= t)

    for i in range(1, n):
        prefix_sum[i][0] = prefix_sum[i-1][0] + int(country[i][0] <= t)

    for j in range(1, m):
        prefix_sum[0][j] = prefix_sum[0][j-1] + int(country[0][j] <= t)

    for i in range(1, n):
        for j in range(1, m):
            prefix_sum[i][j] = int(country[i][j] <= t) + \
                                prefix_sum[i-1][j] + \
                                prefix_sum[i][j-1] - \
                                prefix_sum[i-1][j-1]
    return prefix_sum


def get_sum(prefix_sum: list[list[int]], i: int, j: int, window_size: int) -> int:
    top, left = i - window_size, j - window_size
    sum = prefix_sum[i][j]

    if top >= 0:
        sum -= prefix_sum[top][j]

    if left >= 0:
        sum -= prefix_sum[i][left]

    if top >= 0 and left >= 0:
        sum += prefix_sum[top][left]

    return sum


def is_collapsed(country: list[list[int]], t: int, window_size: int) -> bool:
    n, m = len(country), len(country[0])
    window_size_sq = window_size * window_size
    prefix_sum = fill_prefix_sum(country, t)
    for i in range(window_size - 1, n):
        for j in range(window_size - 1, m):
            if get_sum(prefix_sum, i, j, window_size) == window_size_sq:
                return True
    return False


def main():
    # Lendo os inputs
    n, m, k, _ = map(int, sys.stdin.readline().split())
    country = [[MAX for _ in range(m)] for _ in range(n)]

    for line in sys.stdin:
        a, b, d = map(int, line.split())
        country[a-1][b-1] = d

    # Busca binaria pelo momento em que o pais entrou em colapso
    left, right = 0, MAX
    collapsed = False
    while left < right:
        mid = (right + left) // 2
        if is_collapsed(country, mid, k):
            collapsed = True
            right = mid
        else:
            left = mid + 1

    # Imprime a resposta
    print(right if collapsed else -1)
    

if __name__ == "__main__":
    main()

File: codeforces/MaratonaMineira2022/D/D.cpp
#include <algorithm>
#include <cstdio>
#include <cstring>

#define MAX 1e9 + 1  // Valor maximo de tempo
#define MAX_SIDE 401 // Tamanho maximo de um dos lados do quadrado


void fill_prefix_sum(int country[][MAX_SIDE], int prefix_sum[][MAX_SIDE], int t, int n, int m) {
	memset(prefix_sum, 0, MAX_SIDE * MAX_SIDE * sizeof(int));
	
	prefix_sum[0][0] = (country[0][0] <= t);
	for (int row = 1; row < n; row++)
		prefix_sum[row][0] = prefix_sum[row - 1][0] + (country[row][0] <= t);

	for (int col = 1; col < m; col++)
		prefix_sum[0][col] = prefix_sum[0][col - 1] + (country[0][col] <= t);

	for (int row = 1; row < n; row++) {
		for (int col = 1; col < m; col++) {
			prefix_sum[row][col] = (country[row][col] <= t) + 
						prefix_sum[row-1][col] + 
						prefix_sum[row][col-1] - 
						prefix_sum[row-1][col-1];
		}
	}
}

int get_sum(int prefix_sum[][MAX_SIDE], int i, int j, int window_size){
	int top = i - window_size;
	int left = j - window_size;

	int sum = prefix_sum[i][j];
	if (top >= 0)
		sum -= prefix_sum[top][j];
	if (left >= 0)
		sum -= prefix_sum[i][left];
	if (top >= 0 && left >= 0)
		sum += prefix_sum[top][left];

	return sum;
}

bool is_collapsed(int country[][MAX_SIDE], int t, int n, int m, int window_size){
	int prefix_sum[MAX_SIDE][MAX_SIDE];
	fill_prefix_sum(country, prefix_sum, t, n, m);

	int window_sq = window_size * window_size;
	for (int i = window_size - 1; i < n; i++) {
		for (int j = window_size - 1; j < m; j++) {
			if (get_sum(prefix_sum, i, j, window_size) == window_sq)
				return true;
		}
	}

	return false;
}

int main () {
	int n, m, k, q;
	int a, b, d;

	int country[MAX_SIDE][MAX_SIDE];
	std::fill(&country[0][0], &country[0][0] + MAX_SIDE * MAX_SIDE, MAX);

	// Input dos dados
	scanf("%d %d %d %d", &n, &m, &k, &q);
	for (int i = 0; i < q; i++) {
		scanf("%d %d %d", &a, &b, &d);
		country[a-1][b-1] = d;
	}
	

	// Busca binaria pelo momento em que o pais
	// entrou em colapso
	int mid, l = 0, r = MAX;
	bool collapsed = false;
	while (l < r) {
		mid = (r + l) / 2;
		if (is_collapsed(country, mid, n, m, k)){
			collapsed = true;
			r = mid;
		} else {
			l = mid + 1;
		}

	}

	// Imprime a resposta
	printf("%d\n", (collapsed ? r : -1));

	return 0;
}
